neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt

```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
                 .ifFailedThenRollback(transaction)
     }
 
-    override fun purchasedSubscription(
-            customerId: String,
-            invoiceId: String,
-            chargeId: String,
-            sku: String,
-            amount: Long,
-            currency: String): Either<StoreError, Plan> = writeTransaction {
+    override fun renewSubscriptionToPlan(identity: org.ostelco.prime.model.Identity,
+                                         sku: String): Either<StoreError, Product> = writeTransaction {
+        getCustomer(identity = identity)
+                .flatMap { customer ->
+                    pendingSubscriptionToPlanRelationStore
+                            .get(fromId = customer.id, toId = sku, transaction = transaction)
+                            .flatMap {
+                                paymentProcessor.payInvoice(it.invoiceId)
+                                        .mapLeft {
+                                            NotFoundError(type = "", id = "")
+                                        }
+                            }
+                            .flatMap {
+                                removePendingSubscriptionToPlan(customer.id, sku)
+                            }
+                            .flatMap {
+                                getProduct(identity, sku)   /* Return the subscribed to product. */
+                            }
+                }
+    }
+
+    override fun renewSubscriptionToPlan(identity: org.ostelco.prime.model.Identity,
+                                         sku: String,
+                                         sourceId: String,
+                                         saveCard: Boolean): Either<StoreError, Product> = writeTransaction {
+        getCustomer(identity = identity)
```

Comment:
This is highly unreadable code.  The control flow is obscure, the level of indentation is high.  Please introduce other abstractions (local methods probably) to make the control flow and thus semantics of this statement comprehensible.  Right now it isn't.